File: platforms/android/src/main/java/com/amap/agenui/render/component/impl/IconComponent.java
package com.amap.agenui.render.component.impl;

import android.content.Context;
import android.graphics.Color;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.amap.agenui.render.component.A2UIComponent;
import com.amap.agenui.render.style.IconResourceMapper;
import com.amap.agenui.render.style.StyleHelper;

import java.util.Map;

/**
 * Icon component implementation - compliant with A2UI v0.9 protocol
 *
 * Supported properties:
 * - name: icon name (standard icon or custom path)
 *
 * Standard icons include:
 * accountCircle, add, arrowBack, arrowForward, attachFile, calendarToday, call, camera,
 * check, close, delete, download, edit, event, error, fastForward, favorite, favoriteOff,
 * folder, help, home, info, locationOn, lock, lockOpen, mail, menu, moreVert, moreHoriz,
 * notificationsOff, notifications, pause, payment, person, phone, photo, play, print,
 * refresh, rewind, search, send, settings, share, shoppingCart, skipNext, skipPrevious,
 * star, starHalf, starOff, stop, upload, visibility, visibilityOff, volumeDown,
 * volumeMute, volumeOff, volumeUp, warning
 *
 */
public class IconComponent extends A2UIComponent {

    private Context context;

    private ImageView imageView;

    public IconComponent(Context context, String id, Map<String, Object> properties) {
        super(id, "Icon");
        this.context = context;
        if (properties != null) {
            this.properties.putAll(properties);
        }
    }

    @Override
    protected View onCreateView(Context context) {
        imageView = new ImageView(context);

        // Default icon size 24dp
        int size = (int) (24 * context.getResources().getDisplayMetrics().density);
        imageView.setLayoutParams(new ViewGroup.LayoutParams(size, size));

        // Default color
        imageView.setColorFilter(Color.BLACK);

        // ⚠️ Important: apply initial properties
        onUpdateProperties(this.properties);

        return imageView;
    }

    @Override
    protected void onUpdateProperties(Map<String, Object> changedProps) {
        if (imageView == null) {
            return;
        }

        // Update icon name
        if (changedProps.containsKey("name")) {
            Object nameObj = changedProps.get("name");
            if (nameObj == null) {
                // null (delete signal) → clear icon (align iOS name→nil)
                imageView.setImageDrawable(null);
            } else {
                String iconName = null;

                // name can be a string or an object containing a path
                if (nameObj instanceof String) {
                    iconName = (String) nameObj;
                } else if (nameObj instanceof Map) {
                    Map<String, Object> nameMap = (Map<String, Object>) nameObj;
                    if (nameMap.containsKey("path")) {
                        iconName = String.valueOf(nameMap.get("path"));
                    }
                }

                if (iconName != null) {
                    loadIcon(iconName);
                }
            }
        }

        // Update icon size
        if (changedProps.containsKey("size")) {
            Object sizeObj = changedProps.get("size");
            int sizeDp = 24; // default size

            if (sizeObj instanceof Number) {
                sizeDp = ((Number) sizeObj).intValue();
            } else if (sizeObj instanceof String) {
                try {
                    sizeDp = Integer.parseInt((String) sizeObj);
                } catch (NumberFormatException e) {
                    // use default value
                }
            }

            int sizePx = (int) (sizeDp * context.getResources().getDisplayMetrics().density);
            ViewGroup.LayoutParams params = imageView.getLayoutParams();
            if (params != null) {
                params.width = sizePx;
                params.height = sizePx;
                imageView.setLayoutParams(params);
            }
        }

        // Update icon color
        if (changedProps.containsKey("color")) {
            Object colorObj = changedProps.get("color");
            if (colorObj == null) {
                // null (delete signal) → clear color filter (align iOS color→default)
                imageView.clearColorFilter();
            } else if (colorObj instanceof String) {
                try {
                    int color = Color.parseColor((String) colorObj);
                    imageView.setColorFilter(color);
                } catch (IllegalArgumentException e) {
                    // Color parsing failed, keep default
                }
            }
        }
    }

    /**
     * Load icon.
     * Uses StyleHelper's static method to get the icon resource.
     */
    private void loadIcon(String iconName) {
        // Use StyleHelper's static method to get the standard icon resource ID
        int resId = IconResourceMapper.getIconResourceId(iconName);

        if (resId != 0) {
            imageView.setImageResource(resId);
        } else {
            // Try loading from a custom path
        }
    }
}

File: platforms/ios/AGenUI/Classes/Render/Component/Impl/IconComponent.swift
//
//  IconComponent.swift
//  AGenUI
//
// Created on 2026/2/27.
//

#if AGENUI_SDK_BUILD
import UIKit

/// IconComponent component implementation (compliant with A2UI v0.9 protocol)
///
/// Supported properties:
/// - name: Icon name (custom SVG name or A2UI standard icon name)
/// - color: Icon color (String, supports hex color values)
/// - styles.width: Icon width (via styles field)
/// - styles.height: Icon height (via styles field)
///
/// Design notes:
/// - Icon loading priority: custom SVG icons from Resource/icons directory > SF Symbol fallback
/// - Maps A2UI standard icon names to Lucide Icons filenames (consistent with Android StyleHelper)
/// - Available custom icons: arrow-left/right, bell, calendar, camera, check, heart, star, etc.
/// - Icon reloads on layoutSubviews to match current view size
class IconComponent: Component {
    
    // MARK: - Properties
    
    private var iconImageView: UIImageView?
    private var currentIconName: String?
    private var currentIconSize: CGSize = .zero
    
    // MARK: - Initialization
    
    init(componentId: String, properties: [String: Any]) {
        super.init(componentId: componentId, componentType: "Icon", properties: properties)
        
        // Create icon image view
        let imageView = UIImageView()
        imageView.contentMode = .scaleAspectFit
        imageView.tintColor = .label
        
        addSubview(imageView)
        
        self.iconImageView = imageView
        
        
        // Apply initial properties
        updateProperties(DiffValue.from(properties))
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // MARK: - Measurement Override
    
    override class func measure(type: String, paramJson: String, maxWidth: Float, widthMode: MeasureMode, maxHeight: Float, heightMode: MeasureMode) -> CGSize {
        let defaultSize: CGFloat = 24
        var measuredWidth = defaultSize
        var measuredHeight = defaultSize

        if (widthMode == .exactly || widthMode == .atMost) && maxWidth > 0 {
            measuredWidth = widthMode == .atMost
                ? min(measuredWidth, CGFloat(maxWidth))
                : CGFloat(maxWidth)
        }
        if (heightMode == .exactly || heightMode == .atMost) && maxHeight > 0 {
            measuredHeight = heightMode == .atMost
                ? min(measuredHeight, CGFloat(maxHeight))
                : CGFloat(maxHeight)
        }

        return CGSize(width: measuredWidth, height: measuredHeight)
    }
    
    // MARK: - Component Override
    
    override func updateProperties(_ diff: [String: DiffValue]) {
        // Call parent method to apply CSS properties (e.g., padding, background-color)
        super.updateProperties(diff)
        
        guard let imageView = iconImageView else { return }
        
        // Update icon name (A2UI v0.9 protocol: name)
        if case .value(let nameValue) = diff["name"] {
            let iconName = CSSPropertyParser.extractStringValue(nameValue)
            currentIconName = iconName
            loadIcon(iconName, for: imageView)
        } else if case .deleted = diff["name"] {
            currentIconName = nil
            imageView.image = nil
        }
        
        // Update icon color (A2UI v0.9 protocol: color)
        if case .value(let colorValue) = diff["color"] {
            if let colorString = colorValue as? String {
                imageView.tintColor = parseColor(colorString)
            }
        } else if case .deleted = diff["color"] {
            imageView.tintColor = .label
        }
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        // Sync iconImageView frame with component bounds
        iconImageView?.frame = bounds
        
        // Reload icon to match current size when layout is complete and size is valid
        let currentSize = bounds.size
        guard currentSize.width > 0, currentSize.height > 0,
              currentSize != currentIconSize else { return }
        
        currentIconSize = currentSize
        
        if let iconName = currentIconName, !iconName.isEmpty {
            let mappedName = mapIconNameToLucide(iconName)
            iconImageView?.image = ComponentStyleConfigManager.loadIcon(named: mappedName, size: currentSize)
        }
    }
    
    // MARK: - Private Methods
    
    /// Load icon
    private func loadIcon(_ iconName: String, for imageView: UIImageView) {
        if iconName.isEmpty {
            imageView.image = nil
            return
        }
        
        let mappedName = mapIconNameToLucide(iconName)
        let size = imageView.bounds.size
        if size.width > 0, size.height > 0 {
            imageView.image = ComponentStyleConfigManager.loadIcon(named: mappedName, size: size)
        } else {
            imageView.image = ComponentStyleConfigManager.loadIcon(named: mappedName)
        }
    }
    
    /// Map A2UI standard icon names to Lucide Icons filenames
    /// Mapping is consistent with Android StyleHelper.getIconResourceId()
    private func mapIconNameToLucide(_ iconName: String) -> String {
        switch iconName.lowercased() {
        case "accountcircle":
            return "circle-user"
        case "add":
            return "plus"
        case "arrowback":
            return "arrow-left"
        case "arrowforward":
            return "arrow-right"
        case "attachfile":
            return "paperclip"
        case "calendartoday", "event":
            return "calendar"
        case "call", "phone":
            return "phone"
        case "camera":
            return "camera"
        case "check":
            return "check"
        case "close":
            return "x"
        case "delete":
            return "trash"
        case "download":
            return "download"
        case "edit":
            return "pencil"
        case "error":
            return "circle-alert"
        case "favorite":
            return "heart"
        case "favoriteoff":
            return "heart-off"
        case "folder":
            return "folder"
        case "help":
            return "circle-question-mark"
        case "home":
            return "house"
        case "info":
            return "info"
        case "locationon":
            return "map-pin"
        case "lock":
            return "lock"
        case "lockopen":
            return "lock-open"
        case "mail":
            return "mail"
        case "menu":
            return "menu"
        case "morevert":
            return "ellipsis-vertical"
        case "morehoriz":
            return "ellipsis"
        case "notificationsoff":
            return "bell-off"
        case "notifications":
            return "bell"
        case "payment":
            return "credit-card"
        case "person":
            return "user"
        case "photo":
            return "image"
        case "print":
            return "printer"
        case "refresh":
            return "refresh-cw"
        case "search":
            return "search"
        case "send":
            return "send"
        case "settings":
            return "settings"
        case "share":
            return "share"
        case "shoppingcart":
            return "shopping-cart"
        case "star":
            return "star"
        case "starhalf":
            return "star-half"
        case "staroff":
            return "star-off"
        case "upload":
            return "upload"
        case "visibility":
            return "eye"
        case "visibilityoff":
            return "eye-off"
        case "warning":
            return "triangle-alert"
        default:
            // If no mapping exists, return original name
            return iconName
        }
    }
    
    /// Parse color string
    private func parseColor(_ colorString: String) -> UIColor {
        return UIColor.from(hexString: colorString)
    }
}

#endif // AGENUI_SDK_BUILD
